### DISCUSSION TITLE
Edit-and-continue changes async codegen; make it configurable

### DISCUSSION THREAD

**Discussion by jthorborg** [dotnet/roslyn#51189] [Ideas]
`async`/`await` and "task-like" friends is a really interesting system that actually allows to conjure a completely unmanaged stackless coroutine system, without ever involving the GC. This allows to build performant graph processing systems for games that do not suffer from recursion (and thereby stack overflow) for instance.

Unfortunately, the lowering of the feature changes codegen when running compiling under debug in an effort to support edit and continue - specifically the generated closure is a `class` under debug:
https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncRewriter.cs#L68

This makes it impossible to constrain generic arguments in methods in the builder pointed to by `AsyncMethodBuilder` to be `: unmanaged`.

Worse, lowering happens _after_ proposed source generation and compilation checks _before_ assembly is emitted, so the usual suspects like Cecil cannot be employed to solve this.

I'm working in a domain where edit and continue would never work anyway, so a reasonable compromise could be having a compiler flag for overriding whether EnC is enabled, looks like `CompilationOptions` was already encapsulating this perhaps with that in mind?

Thanks for any thoughts!

**Comment by tmat** [comment]
Could you share a self-contained code snippet that you'd like to compile that does not today?

**Reply by jthorborg** [reply]
Here you go, [also on sharplab](https://sharplab.io/#v2:D4AQTAjAsAUCAMACEEB0AlArgOwC4EsBbAU1QEk9iAnAewAcBlagN3wGNiBnAblgWTRY8RUgGEahOvgA21JlVYcesWAG0GuKpja4AMgEMAnjUy4AFAeOmA0vmwATVEwCOmYsP3SAlAF0+AZkROTW1cRAB5OgAhTBl7agAeABUAPlhEDMQAdwALamJEJMQALkQcQn1sfQBzYntYAG90zPQTAmxiZJTEQgB9dGJXLlxeGGaMkECUADYy7E59ADMCyJi4xNTEUSpifVxiMy9EAF5u+MX9TGkRlRhM5EDW0ztOzZBZ057+wbdg0fGHsgACyIJi4AacK7mIo7SHXThHBqIAC+t3ukzmC2WwMQAEEsvp8LhwthxJJZPt7Ml8YT9lQADSFDR7YgAWX0bByLxSZh2i0KNKJ1GQAE5qIy+UzcCz2ZyXkFpftZVyOl4AfdcvkBQShVQSnMKlVavZGWQAHI0AiLQxkugU/A0bDqzKanZSmUclUFUrlSo1Oqm3GcQzYNjMpWel4App3TIA1FjWMTQI4LEFEAg5lUXDJcNsyMdHmSpJ55Xy4IeuWq50ZV0FEuK/NV70Gv3GwPB0OlgvEaNozIJgEYjOg4i4ACiAA8OHQCI6zFOZ3PsIhiIiUf3kzjBbgAKrzJbEEm2il1ak6umMhuVr1F4j8pI74UgMUMxDF7vNhU3l5qpP3bI8jdR8L2FUoyG2Il2E8C0rRtCQ7THB0nX/DUgPrT8vX1MggxDMNGzLDo+3/BN0SmTMx0wl4zBwzt8J/DpvwjZsjk+XAcloLJEA6LjYLIcliBISh7EXYhZ2Qw5RkHGBVFw0NWTHHIaHsNZpHiKgzFwQw6GIGhFjMVZYjUxIUi8Xw1A0LQdEsNoLCMNpbAcJwfncAhPHMuAUwPbFgissInnaV40n/OtCn1X0jTqRoh0CJ89QAcTHOLDhOboeLxUCNPY/AEX+f8MV80IMtpMDEHNS18GtE8kMdYiAJaNoXi6L4BiGP5NwAjFkoCprNmlTgAGtdHwAa1xrRAY3q+q+la34wmORB+qGkbiFGKaN0TdaMSKRLwS4KEUs+c5LmuPKtsCAAjGgaGkMrOGqylUsQY6oTOqbhxBY8ENPewzBQMBEDYR12kwPZkNY7p2M47jiF4y1+MQoTKVE8T5y8NaB1gaThwB0RovyqYAA5EB6joEjsXBunZOxDjq96AHZEAAVgxjJSLjGBkSAA=):

```
using System;
using System.Runtime.InteropServices;
using System.Runtime.CompilerServices;

[StructLayout(LayoutKind.Sequential)]
public struct OpBuilder<T>
    where T : unmanaged
{
    Routine<T> m_Request;

    public static unsafe OpBuilder<T> Create() => default;

    public Routine<T> Task => m_Request;

    public void SetResult(T results) { }

    public unsafe void AwaitOnCompleted<TAwaiter, TStateMachine>(ref TAwaiter awaiter, ref TStateMachine stateMachine)
        where TAwaiter : unmanaged, INotifyCompletion
        where TStateMachine : unmanaged, IAsyncStateMachine
    {
    
    }

    public unsafe void Start<TStateMachine>(ref TStateMachine stateMachine)
        where TStateMachine : unmanaged, IAsyncStateMachine
    {

    }

    public void SetException(Exception e) { }

    public void AwaitUnsafeOnCompleted<TAwaiter, TStateMachine>(ref TAwaiter awaiter, ref TStateMachine stateMachine)
        where TAwaiter : ICriticalNotifyCompletion
        where TStateMachine : IAsyncStateMachine
    {
    }
    public void SetStateMachine(IAsyncStateMachine stateMachine) => throw new NotImplementedException();
}

[AsyncMethodBuilder(typeof(OpBuilder<>))]
[StructLayout(LayoutKind.Sequential)]
public unsafe struct Routine<T>
    where T : unmanaged
{
    public Awaiter GetAwaiter() => new Awaiter(this);

    public struct Awaiter : INotifyCompletion
    {
        Routine<T> m_Request;

        public Awaiter(Routine<T> taskLike)
        {
            m_Request = taskLike;
        }

        public T GetResult() => default;

        public bool IsCompleted => default;

        public void OnCompleted(Action continuation) => throw new NotImplementedException();
    }
}

public class C
{
    public async Routine<int> Main()
    {
        return 5;
    }
    
}
```

It works fine in release but not in debug.

**Reply by tmat** [reply]
Filed an issue: https://github.com/dotnet/roslyn/issues/51234

**Reply by tmat** [reply]
The issue describes the necessary fix. If you're interested in fixing this we'd accept a PR.

**Reply by jthorborg** [reply]
Thanks for writing that up, I think it's a better approach than what I suggested for sure.
Changes to async in the compiler seems straightforward, and I'll take a look to see if it's something I can reasonably whip up. I'm not sure about integration about the IDE and how that works though, perhaps I'll stumble upon that during investigation.
